src/main/java/com/puppycrawl/tools/checkstyle/MultiThreadChecker.java

```diff
@@ -0,0 +1,226 @@
+            }
+        }
+    }
+
+    /**
+     * Runs all the given tasks.
+     * @param tasks A list of tasks to run.
+     * @throws InterruptedException see {@link MultiThreadChecker#finishTask(Future)}.
+     * @throws CheckstyleException see {@link MultiThreadChecker#finishTask(Future)}.
+     * @throws ExecutionException see {@link MultiThreadChecker#finishTask(Future)}.
+     */
+    private void runAllTasks(List<Callable<ChecksExecutionResult>> tasks)
+            throws InterruptedException, CheckstyleException, ExecutionException {
+
+        final List<Future<ChecksExecutionResult>> futures = tasks.stream()
+                .map(task -> executor.submit(task))
+                .collect(Collectors.toList());
+
+        for (Future<ChecksExecutionResult> future : futures) {
+            finishTask(future);
+        }
+    }
+
+    /**
+     * Finishes the given future.
+     * @param future A future to finish.
+     * @throws InterruptedException might be thrown from {@link Future#get()}.
+     * @throws CheckstyleException might be thrown from a task.
+     * @throws ExecutionException might be thrown from {@link Future#get()}.
+     */
+    private static void finishTask(Future<ChecksExecutionResult> future)
+            throws InterruptedException, CheckstyleException, ExecutionException {
+
+        final ChecksExecutionResult result = future.get();
+        if (result.getThrownException() != null) {
+            throw result.getThrownException();
+        }
+    }
+
+    /**
+     * Creates a callable for processing a particular file.
+     * @param file A file to process.
+     * @return A callable which processes the given file.
+     */
+    private Callable<ChecksExecutionResult> createProcessFileCallable(File file) {
+
+        final String fileName = file.getAbsolutePath();
+
+        return () -> {
+            ChecksExecutionResult result;
+            try {
+                final SortedSet<LocalizedMessage> messages = processFile(file);
+                fireErrors(fileName, messages);
+                fireFileFinished(fileName);
+                result = ChecksExecutionResult.getCompleted(file, messages);
+            }
+            catch (CheckstyleException ex) {
+                result = ChecksExecutionResult.getFailed(file, ex);
+            }
+
+            return result;
+        };
+    }
+
+    @Override
+    protected SortedSet<LocalizedMessage> processFile(File file) throws CheckstyleException {
+        List<FileSetCheck> clones = threadLocalFilesetCheckClones.get();
+        if (clones == null) {
+            clones = CheckCloneService.cloneFileSetChecks(getFileSetChecks());
+            threadLocalFilesetCheckClones.set(clones);
+            fileSetCheckClones.addAll(clones);
+        }
+
+        return processFile(file, clones);
+    }
+
+    /**
+     * A check execution result.
+     * @author Andrew Kuchev.
+     */
+    private static final class ChecksExecutionResult {
```

Comment:
Expand documentation.
Checker calls filesets, not checks. So shouldn't this be called `CheckerExecutionResults`?